src/shared/services/s3.service.ts

```diff
@@ -0,0 +1,136 @@
+import {
+  DeleteObjectCommand,
+  GetObjectCommand,
+  HeadObjectCommand,
+  PutObjectCommand,
+  S3Client,
+} from '@aws-sdk/client-s3'
+import { getSignedUrl } from '@aws-sdk/s3-request-presigner'
+import {
+  Injectable,
+  InternalServerErrorException,
+  NotFoundException,
+} from '@nestjs/common'
+import { ConfigService } from '@nestjs/config'
+import type {
+  IBaseS3,
+  IResponseMessage,
+  IResponseUploadUrl,
+  IUploadUrl,
+} from '@shared/domain/S3Type'
+import type { IS3Service } from '@shared/domain/services/IS3Service'
+
+@Injectable()
+export class S3Service implements IS3Service {
+  private client: S3Client
+  private bucketName: string
+  private expiration: number
+
+  constructor(private readonly configService: ConfigService) {
+    const s3Region = this.configService.get<string>('S3_REGION')
+    const accessKeyId = this.configService.get<string>('S3_ACCESS_KEY')
+    const secretAccessKey = this.configService.get<string>(
+      'S3_SECRET_ACCESS_KEY',
+    )
+    const expiration = Number(this.configService.get<string>('EXPIRATION'))
+    this.bucketName = this.configService.get<string>('S3_BUCKET_NAME')
+
+    if (
+      !s3Region ||
+      !accessKeyId ||
+      !secretAccessKey ||
+      !this.bucketName ||
+      !expiration
+    ) {
+      throw new Error('Missing S3 configuration in environment variables')
+    }
+
+    this.client = new S3Client({
+      region: s3Region,
+      credentials: {
+        accessKeyId,
+        secretAccessKey,
+      },
+    })
+  }
+
+  async generatePresignedUploadUrl({
+    fileType,
+    fileName,
+  }: IUploadUrl): Promise<IResponseUploadUrl> {
+    try {
+      const mimeType =
+        fileType === 'mp4'
+          ? 'video/mp4'
+          : fileType === 'png'
+            ? 'image/png'
+            : 'image/jpg'
+
+      const key = `${uuidv4()}`
+
+      const command = new PutObjectCommand({
+        Bucket: this.bucketName,
+        Key: key,
+        ContentType: mimeType,
+        ACL: 'public-read',
+        Metadata: {
+          fileName,
+        },
+      })
+
+      const url = await getSignedUrl(this.client, command, {
+        expiresIn: this.expiration,
+      })
+      return { url, key }
+    } catch (error) {
+      throw new InternalServerErrorException(
+        'Error generating presigned upload URL',
+      )
+    }
+  }
+
+  async generatePresignedGetUrl({ key }: IBaseS3): Promise<string> {
+    try {
+      const command = new GetObjectCommand({
+        Bucket: this.bucketName,
+        Key: key,
+      })
+
+      return await getSignedUrl(this.client, command, { expiresIn: 3600 })
+    } catch (error) {
+      throw new InternalServerErrorException(
+        'Error generating presigned get URL',
+      )
+    }
+  }
+
+  async deleteFile({ key }: IBaseS3): Promise<IResponseMessage> {
+    try {
+      await this.fileExist(key)
+      const command = new DeleteObjectCommand({
+        Bucket: this.bucketName,
+        Key: key,
```

Comment:
esto no va aqui:
 await this.fileExist(key)
 dejalo separado, esa validacion se hace en el service donde es invocado y se hace por separado, esto es porque cada funcion debe tener una responsabilidad unica, ahi ya le cargaste dos